rfcs/0002-branch-generation.md

```diff
@@ -0,0 +1,76 @@
+
+
+## Basic Design Idea
+
+Nodes take turns being leader and generating the PoH that encodes state changes.  The network can tolerate loss of connection to any leader by synthesizing what the leader ***would have generated*** had it been connected but not ingesting any state changes.  The complexity of forks is thereby limited to a "there/not-there" skip list of branches that may arise on leader rotation periods boundaries.  A leader can only transmit durring their predefined PoH slot.
+
+## Message Flow
+
+1. Transactions are ingested at the current leader.
+2. Leader filters for valid transactions.
+3. Leader executes valid transactions on its state.
+4. Leader packages transactions into entries based off its current PoH slot.
+5. Leader transmits the entries to validator nodes (in signed blobs)
+    a. The PoH stream includes ticks; empty entries that indicate liveness of the leader and the passage of time on the network.
+    b. A leader's stream begins with the tick entries necessary complete the PoH back to the leaders most recently observed prior leader period.
+6. Validators retransmit entries to peers in their set and to further downstream nodes.
+7. Validators validate the transactions and execute them on their state.
+8. Validators compute the hash of the state.
+9. At specific times, i.e. specific PoH tick counts, validators transmit votes to the leader.
+    a. Votes are signatures of the hash of the computed state at that PoH tick count
+    b. Votes are also propagated via gossip
+10. Leader executes the votes as any other transaction and broadcasts them to the network
+11. Validators observe their votes and all the votes from the network.
+
+## Partitions, Forks
+
+Forks can arise at PoH tick counts that correspond to a vote.  The next leader may not have observed the last vote period and may start their slot with generated virtual PoH entries.  These empty ticks are generated by all nodes in the network at a network-specified rate for hashes/per/tick `Z`.
+
+There are only two possible versions of the PoH during a voting period: PoH with `T` ticks and entries generated by the current leader, or PoH with just ticks.  The "just ticks" version of the PoH can be thought of as a virtual ledger, one that all nodes in the network can derive from the last tick in the previous period.
+
+Validators can ignore forks at other points (e.g. from the wrong leader), or slash the leader responsible for the fork.
+
+Validators vote based on a greedy choise to maximze their reward described in [branch selection](rfcs/0008-branch_selection.md).
+
+### Validator's View
+
+#### Time Progression
+The diagram below represents a validator's view of the PoH stream with possible forks over time.  L1, L2, etc. are leader periods, and `E`s represent entries from that leader during that leader's period.  The 'x's represent ticks only, and time flows downwards in the diagram.
+
+
+```
+time  +----+                                          validator action
```

Comment:
Can you copy in `book/art/leader-scheduler.bob` so that it's easy for me to see any changes you make to it?  Those parentheses don't render well, so I change `E(L1)` to `E1`. Seems to be just as meaningful and renders nicely by svgbob.